python/cuml/internals/array.py

```diff
@@ -255,7 +255,7 @@ def __init__(self,
                     self._mem_type = GlobalSettings().memory_type
 
                 try:
-                    data = data.ptr
+                    data = data.get_ptr()
```

Comment:
This is correct (cuml is taking ownership of the buffer, so we can't do anything), but I _think_ this is an impossible branch? How could we get here given that if data is a `Buffer` that exposes `get_ptr` then it also has a `__cuda_array_interface__` property so we'll never end up in this branch?

Reply:
You are right, we will never branch for a `Buffer` to this part of the code. Reverted the change.

Reply:
Can a cuml-ite cast some light on what is meant to be going on in this function? I _think_ that this patch:
```patch
diff --git a/python/cuml/internals/array.py b/python/cuml/internals/array.py
index 808b92fe3..f9386ee8a 100644
--- a/python/cuml/internals/array.py
+++ b/python/cuml/internals/array.py
@@ -242,9 +242,7 @@ class CumlArray():
                             'Must specify dtype when data is passed as a'
                             ' {}'.format(type(data))
                         )
-                if isinstance(data, (CudfBuffer, DeviceBuffer)):
-                    self._mem_type = MemoryType.device
-                elif mem_type is None:
+                if mem_type is None:
                     if GlobalSettings().memory_type in (
                         None, MemoryType.mirror
                     ):
@@ -254,13 +252,6 @@ class CumlArray():
                         )
                     self._mem_type = GlobalSettings().memory_type
 
-                try:
-                    data = data.ptr
-                    if shape is None:
-                        shape = (data.size,)
-                    self._owner = data
-                except AttributeError:  # Not a buffer object
-                    pass
                 if isinstance(data, int):
                     self._owner = owner
                 else:
```
Just deletes unreachable code:

- Both `CudfBuffer` and `DeviceBuffer` advertise a `__cuda_array_interface__` property
- Only `CudfBuffer` and `DeviceBuffer` have a `.ptr` property

So if `data` is either a `CudfBuffer` or a `DeviceBuffer` then we will never fall through the initial

```
try:
    data.__cuda_array_interface__
except AttributeError:
   ...
```
check.

@wphicks can you comment?

Reply:
@wence- Yep, that patch looks right to me. This code was based on an earlier version that did not take advantage of `__cuda_array_interface__` at all. To answer the question about what this function is meant to, it helps construct a unified interface for all the data types cuML accepts. This function harvests all the data necessary for the CumlArray to populate a `__cuda_array_interface__` or `__array_interface__` of its own, even if the original object does not have an array interface.

Reply:
thanks, I'll submit something as a followup, in the interests of keeping Prem's change as small as possible

Reply:
https://github.com/rapidsai/cuml/pull/5166